### PR TITLE
Feature/holger/reducesum atomic

### PR DESCRIPTION
**Pull request by jonesholger**
Added a ReduceSum variant driven by a new cuda_reduce_atomic policy switch. This code is similar to the min_migration variants which use atomics to update a small tally block. The reducesum unit test #3 dsum2 exercises this variant.

For discussion is whether to create a whole new series of tests which would try to expose non-determinism (non-trivial), or to rename the policy cuda_reducesum_atomic (benign)

### CODE REVIEW

File: include/RAJA/exec-cuda/reduce_cuda.hxx [hunk 1]
```diff
@@ -308,58 +333,44 @@ template <size_t BLOCK_SIZE, typename T>
 class ReduceMin< cuda_reduce<BLOCK_SIZE>, T > 
 {
 public:
```
Comment (keasler) [inline include/RAJA/exec-cuda/reduce_cuda.hxx]: I am super-confused.... isn't this the third time I have approved these implementations of ReduceMin and ReduceMax?  Why do they keep showing up? @davidbeckingsale I'd like to get a handle on why this has shown up in three separate commits.

Comment (jonesholger) [inline include/RAJA/exec-cuda/reduce_cuda.hxx]: Probably my fault, I should do a rebase as this PR is stale and you're seeing weird commit changes. I thought the diffs were smarter (i.e. diff'ed against "current" develop)

File: include/RAJA/exec-cuda/reduce_cuda.hxx [hunk 2]
```diff
@@ -607,13 +571,7 @@ public:
...
       if (threadIdx.x < 1) {
           sd[0] = RAJA_MAX(sd[0],sd[1]);
-#if defined(RAJA_USE_NO_ATOMICS)
-          m_blockdata[m_blockoffset + blockIdx.x+1]  =
-              RAJA_MAX( sd[0], m_blockdata[m_blockoffset + blockIdx.x+1] );
-
-#else
-          atomicMax( &m_blockdata[m_blockoffset], sd[0] );
-#endif
+          atomicMax(&(m_tallydata->tally), sd[0] );
```
Comment (keasler) [inline include/RAJA/exec-cuda/reduce_cuda.hxx]: Note that the argument to this class is cuda_reduce<>, and not cuda_reduce_atomic<>.  The ReduceMin class implemented immediately above this class is implemented correctly.  This class should look like the class immediately above it.

**Comment by keasler** [comment]
I approve this branch. @jonesholger 

 I notice that I did not have to re-review the minloc and maxloc, but that I did have to re-review the min and max.  I hate git.  it is very confusing why I have to review some parts of the code twice -- git should be smart enough to see 'true' differences vs the develop branch. @davidbeckingsale

**Comment by davidbeckingsale** [comment]
@keasler stop spamming me. The reason this will show up is due to when @jonesholger created the branches, and the points in git history that share common commits.

**Comment by rhornung67** [comment]
@jonesholger is there an easy way to initialize the minloc/maxloc reduction stucts (reduced value and index) on the device like the others? I can't comment on a specific line number here since those lines are not changed in this PR.

**Comment by jonesholger** [comment]
@rhornung67 those classes already have initializers.

from the unit test

ReduceMaxLoc< cuda_reduce<block_size>, double> dmax0(-DBL_MAX,-1);

and in the class
explicit ReduceMaxLoc(T init_val, Index_type init_loc)
 954    {
 955       m_is_copy = false;
 956       m_reduced_val = init_val;
 957       m_reduced_idx = init_loc;
 958       m_myID = getCudaReductionId();
 959       retiredBlocks[m_myID] = 0;
 960       m_blockdata = getCudaReductionLocMemBlock(m_myID) ;
 961       m_blockoffset = 1;
 962       m_blockdata[m_blockoffset].val = init_val;
 963       m_blockdata[m_blockoffset].idx = init_loc;
 964 
 965       for (int j = 1; j <= RAJA_CUDA_REDUCE_BLOCK_LENGTH; ++j) {
 966          m_blockdata[m_blockoffset+j].val = init_val;
 967          m_blockdata[m_blockoffset+j].idx = init_loc;
 968 
 969       }
 970       cudaErrchk(cudaDeviceSynchronize());
 971    }
 972

**Pull request merged**